utils/sort_db.py

```diff
@@ -0,0 +1,352 @@
+import sys
+
+import json
+import xjson
+import cmp
+
+
+def split_all(s, chars):
+    """Split on multiple character values.
+
+    >>> split_all('a_b_c_d', '_. ')
+    ['a', 'b', 'c', 'd']
+    >>> split_all('a b c d', '_. ')
+    ['a', 'b', 'c', 'd']
+    >>> split_all('a.b.c.d', '_. ')
+    ['a', 'b', 'c', 'd']
+    >>> split_all('a_b.c d', '_. ')
+    ['a', 'b', 'c', 'd']
+    >>> split_all('a b_c.d', '_. ')
+    ['a', 'b', 'c', 'd']
+    """
+    chars = list(chars)
+
+    o = [s]
+    while len(chars) > 0:
+        c = chars.pop(0)
+
+        n = []
+        for i in o:
+            n += i.split(c)
+
+        o = n
+    return o
+
+
+NUM_REGEX = re.compile('^(.*?)([0-9]*)$')
+
+
+def extract_num(i):
+    """Extract number from a string to be sorted.
+
+    >>> extract_num('BLAH123')
+    ('BLAH', 123)
+    >>> extract_num('123')
+    123
+    >>> extract_num('BLAH')
+    'BLAH'
+    >>> extract_num('')
+    ''
+    """
+    g = NUM_REGEX.match(i).groups()
+    if len(g[-1]) == 0:
+        return i
+    i = int(g[-1])
+    if len(g[0]) == 0:
+        return i
+    else:
+        return (g[0], i)
+
+
+class bit(tuple):
+    """Class representing a bit specifier.
+
+    >>> a = bit.parse("02_12")
+    >>> a
+    bit(2, 12, True)
+    >>> b = bit.parse("!02_03")
+    >>> b
+    bit(2, 3, False)
+    >>> b == a
+    False
+    >>> b < a
+    True
+    >>> str(a)
+    '02_12'
+    >>> str(b)
+    '!02_03'
+
+    >>> bit.parseline("!30_04 !31_00 !31_01 31_02")
+    [bit(30, 4, False), bit(31, 0, False), bit(31, 1, False), bit(31, 2, True)]
+
+    >>> bit.parseline("31_02 !31_00 !31_01 !30_04")
+    [bit(30, 4, False), bit(31, 0, False), bit(31, 1, False), bit(31, 2, True)]
+    """
+
+    @classmethod
+    def parse(cls, s):
+        mode = s[0] != '!'
+        s = s.replace('!', '')
+        assert '_' in s, s
+        a, b = s.split('_', 1)
+        assert '_' not in b, s
+        return cls((extract_num(a), extract_num(b), mode))
+
+    @classmethod
+    def parseline(cls, s):
+        bits = [cls.parse(b) for b in s.split(' ')]
+        bits.sort()
+        return bits
+
+    def __repr__(self):
+        return "bit" + tuple.__repr__(self)
+
+    def __str__(self):
+        s = self
+        return "{}{:02d}_{:02d}".format(['!', ''][s[2]], s[0], s[1])
+
+
+def convert_bit(i):
+    """Convert a bit pattern into sortable form.
+
+    >>> convert_bit("02_12")
+    bit(2, 12, True)
+    >>> convert_bit("!02_12")
+    bit(2, 12, False)
+    >>> convert_bit("!02_02")
+    bit(2, 2, False)
+    >>> convert_bit("always")
+    'always'
+    """
+    if '_' not in i:
+        return i
+    return bit.parse(i)
+
+
+def segbit_line_sort_bits(l):
+    """Sort the bit section of a segbit line.
+
+    >>> segbit_line_sort_bits("A !28_35 !27_39 27_37")
+    'A 27_37 !27_39 !28_35'
+
+    >>> segbit_line_sort_bits("B !28_35 !27_39 !27_37")
+    'B !27_37 !27_39 !28_35'
+
+    >>> segbit_line_sort_bits("C 28_35 00_00 !27_37")
+    'C 00_00 !27_37 28_35'
```

Comment:
No, put negative bits first or last.  Do not intermix.

Reply:
See reason in docstring of whole module.

Reply:
But 1 hot encoding are relatively rare.  The bits that are set are more important than the bits that are cleared.

Reply:
It also makes it hard to see when a diff has a bit flip.

Reply:
Also, when you sort this way, a number of physical structures become much clearer (specially around chained muxes).